### ISSUE TITLE
Discussion on some style topics

### ISSUE THREAD

**Issue by dburriss** [CompositionalIT/farmer#127]
This issue is a small catch-all for a few things I noticed diving into the codebase. I can break it into separate issues if any are valid after discussion, as it may just be ignorance on my side.

**1. Sln / folder structure mismatch**
Currently, there is a mismatch between the folder structure of the solution and what is shown in an IDE Solution explorer.

_Folder view:_
/src/Farmer/
/src/SampleApp/
/src/Samples/
/src/Tests/

_Solution view:_
src/Tests
Farmer
SampleApp

I am not sure if this was on purpose or the consequence of working mostly in folder view of VS Code?

Suggestion

_Folder view:_
/src/Farmer/
/samples/SampleApp/
/samples/Scripts/
/src/Tests/   (or /test/Farmer ? ... honestly I don't think /test/ folder is necessary since this is a very small project)

_Solution view:_
Tests
Farmer
samples/SampleApp
samples/ _script links_

The idea on the solution view here is it keeps the 2 projects that are touched most shallow in the structure. It is a small library with few projects, the simpler the better.

**2. `AutoOpen` on the features means common types like Sku cases are subject to compiled last**

Since `[<AutoOpen>]` is on the Builder files, even when using a single resource you could not use a Sku of `Basic` unless by chance it was the resource for the last builder in the project. When Basic shows up in Intellisense but then does not work it is a little unexpected. It would be more expected if I explicitly opened a resource namespace.

**3. Test style**
Since the above points are largely about the style I thought I would mention it may be worth noting (although not do anything about it now) that in the future some guidelines on test style should be given (if contributor numbers increase). I looked and there were 2 distinct styles, so I added a 3rd :) 

Mine could use some work so not married to them. They do successfully follow a few tenants I have for tests though.
  - Clear Setup: Make it clear what the setup inputs are
  - Readable: Make it easy to read what the test is testing
  - Declarative: Declare what you want without tying each test to implementation details
  - Outside-In: Unless working on something specific, test from the outside so you are not tying your tests to internal implementation details

**4. Script files are tailored for running on Windows only**
The references to DLLs use `\` instead of `/`.

**Comment by isaacabraham** [comment]
All valid points @dburriss. Taking them one at a time:

1. The solution file needs a clean up - happy for you to create an issue and fix it as required. The folder is probably only there for tests because I used `dotnet sln` to add it rather than doing it manually.

**Action**: Create issue, PR if possible :-)

2. The SKUs are an issue that I've noticed too. I'm tempted to simply put `[<RequiredQualifiedAccess>]` on all of them and ensure that they all have different type names e.g. `StorageSku` instead of just `Sku`. What do you think? The `[<AutoOpen>]` was designed to aid discoverability insofar as people can just open Farmer.Resources and get access to everything. Maybe that was a mistake.

**Action**: Create issue for ongoing discussion?

3. I'm not too fussed about the style - consistency is great to have but difficult in OSS projects :-) The DSL style you have is very nice and readable although the cost of creating that for a single use feels high to me, so I probably wouldn't do it myself, but I'm not averse to it :-)

4. Issue / PR fix welcome.

**Comment by dburriss** [comment]
@isaacabraham I will create issues for points 1,2,4.

WRT 3. Honestly, I am not too fussed about the consistency either. As I said, I wasn't suggesting even doing anything to fix. I will reiterate something I did mention and that was adding a testing section to contributions. This wouldn't focus too much on Arrange or Assert style but more granularity. What we want (based on existing tests and my preference too) is outside-in black-box style tests that validate against the authors knowledge of the ARM template, not against Azure. If there are lower level tests of _converters_ or _outputters_ they need to be in addition to outside-in tests. This makes it easier rather than harder to contribute in my mind since it gives guidance. I had to go dig and divine what the intent was. 
So I could add something about this is the contributions markdown mentioned in #128 ?

With respect to the effort for the declarative assertions... I agree it is too much effort to be a recommended practice for this project (or most). Since I don't really get to code for work these days, I need to experiment, see what works, and have some fun somewhere :P
